packages/ledgerstate/color.go

```diff
@@ -0,0 +1,228 @@
+	case ColorIOTA:
+		return "IOTA"
+	case ColorMint:
+		return "MINT"
+	default:
+		return c.Base58()
+	}
+}
+
+// ColorIOTA is the zero value of the Color and represents uncolored tokens.
+var ColorIOTA = Color{}
+
+// ColorMint represents a placeholder Color that indicates that tokens should be "colored" in their Output.
+var ColorMint = Color{255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255}
+
+// endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// region ColoredBalances //////////////////////////////////////////////////////////////////////////////////////////////
+
+// ColoredBalances represents a collection of balances associated to their respective Color that maintains a
+// deterministic order of the present Colors.
+type ColoredBalances struct {
+	balances *orderedmap.OrderedMap
+}
+
+// NewColoredBalances returns a new deterministically ordered collection of ColoredBalances.
+func NewColoredBalances(balances map[Color]uint64) (coloredBalances *ColoredBalances) {
+	coloredBalances = &ColoredBalances{balances: orderedmap.New()}
+
+	// deterministically sort colors
+	sortedColors := make([]Color, 0, len(balances))
+	for color := range balances {
+		sortedColors = append(sortedColors, color)
+	}
+	sort.Slice(sortedColors, func(i, j int) bool { return bytes.Compare(sortedColors[i][:], sortedColors[j][:]) < 0 })
+
+	// add sorted colors to the underlying map
+	for _, color := range sortedColors {
+		coloredBalances.balances.Set(color, balances[color])
+	}
+
+	return
+}
+
+// ColoredBalancesFromBytes unmarshals ColoredBalances from a sequence of bytes.
+func ColoredBalancesFromBytes(bytes []byte) (coloredBalances *ColoredBalances, consumedBytes int, err error) {
+	marshalUtil := marshalutil.New(bytes)
+	if coloredBalances, err = ColoredBalancesFromMarshalUtil(marshalUtil); err != nil {
+		err = xerrors.Errorf("failed to parse ColoredBalances from MarshalUtil: %w", err)
+		return
+	}
+	consumedBytes = marshalUtil.ReadOffset()
+
+	return
+}
+
+// ColoredBalancesFromMarshalUtil unmarshals ColoredBalances using a MarshalUtil (for easier unmarshaling).
+func ColoredBalancesFromMarshalUtil(marshalUtil *marshalutil.MarshalUtil) (coloredBalances *ColoredBalances, err error) {
+	balancesCount, err := marshalUtil.ReadUint32()
+	if err != nil {
+		err = xerrors.Errorf("failed to parse element count (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+
+	var previousColor *Color
+	coloredBalances = NewColoredBalances(nil)
+	for i := uint32(0); i < balancesCount; i++ {
+		color, colorErr := ColorFromMarshalUtil(marshalUtil)
+		if colorErr != nil {
+			err = xerrors.Errorf("failed to parse Color from MarshalUtil: %w", colorErr)
+			return
+		}
+
+		// check semantic correctness (ensure ordering)
+		if previousColor != nil && bytes.Compare(previousColor[:], color[:]) != -1 {
```

Comment:
If `Color` has a method to compare then this could be simplified to be more readable.

Reply:
agree